### PR TITLE
SolveParams class and separate solve function

### PR DESCRIPTION
**Pull request by leonlan**
This PR introduces a SolveParams class for solver parameter configuration and separates the `solve` function, which both Model and CLI use. This is useful for addressing #454.

### CODE REVIEW

**Comment by leonlan** [comment]
@N-Wouda Can you give me your opinion on this?

File: pyvrp/search/__init__.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 from .neighbourhood import NeighbourhoodParams as NeighbourhoodParams
 from .neighbourhood import compute_neighbours as compute_neighbours
 
-NODE_OPERATORS = [
+NODE_OPERATORS: list[NodeOperator] = [
```
Comment (leonlan) [inline pyvrp/search/__init__.py:20]: Mypy complains about this but I don't understand why.

Comment (N-Wouda) [reply pyvrp/search/__init__.py:20]: These are classes, not instantiated objects. This should work: `list[Type[NodeOperator]]` (using `typing.Type`).

File: pyvrp/ParamConfig.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+from pyvrp.search import (
+    NODE_OPERATORS,
+    ROUTE_OPERATORS,
+    NeighbourhoodParams,
+    NodeOperator,
+    RouteOperator,
+)
+
+
+class ParamConfig:
```
[the remark this answers is not in the document]
Comment (N-Wouda) [reply pyvrp/ParamConfig.py:19]: Not sure about the name, let's discuss that a bit more.

**Comment by leonlan** [comment]
Will continue working on this later this week!

File: pyvrp/cli.py
[the remark this answers is not in the document]
Comment (leonlan) [reply pyvrp/cli.py:1]: It takes about 2 seconds to construct a 1000-client instance.

[the remark this answers is not in the document]
Comment (N-Wouda) [reply pyvrp/cli.py:1]: Ah I assumed `MaxRuntime`'s clock started ticking the moment it got constructed, but that's not the case. This should be OK then.

File: pyvrp/cpp/SubPopulation.h
Comment (N-Wouda) [inline pyvrp/cpp/SubPopulation.h:1]: The whole `PopulationParams` needs some love, which I'll try to get around to sometime soon.

File: pyvrp/cli.py [hunk 3]
```diff
@@ -310,6 +254,9 @@ def main():
     msg = "Whether to scale stopping criteria values by the number of clients."
     stop.add_argument("--per_client", action="store_true")
 
+    msg = "Whether to display information about the solver progress."
+    parser.add_argument("--display", action="store_true", help=msg)
```
Comment (N-Wouda) [inline pyvrp/cli.py:258]: Does this work without passing the `stats_dir` argument as well? And how does this interact with the progress bar when solving a set of instances? I am unsure if we should support this from the CLI.

If we do, the argument should probably be added above the `stop` argument group. It's much easier to separate the different parts of the CLI that way.

File: docs/source/conf.py [hunk 4]
```diff
@@ -30,6 +30,7 @@
 autoclass_content = "class"
 autodoc_member_order = "bysource"
 autodoc_typehints = "signature"
+autodoc_preserve_defaults = True
```
Comment (N-Wouda) [inline docs/source/conf.py:33]: Looks OK. I didn't see anything weird when browsing the API docs!

File: pyvrp/SolveParams.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        """
+        with open(loc, "rb") as fh:
+            data = tomli.load(fh)
+
+        gen_params = GeneticAlgorithmParams(**data.get("genetic", {}))
+        pen_params = PenaltyParams(**data.get("penalty", {}))
+        pop_params = PopulationParams(**data.get("population", {}))
+        nb_params = NeighbourhoodParams(**data.get("neighbourhood", {}))
+        node_ops = [getattr(pyvrp.search, op) for op in data["node_ops"]]
+        route_ops = [getattr(pyvrp.search, op) for op in data["route_ops"]]
```
Comment (N-Wouda) [inline pyvrp/SolveParams.py:92]: If not passed, should we then not default to `NODE_OPERATORS`, or `ROUTE_OPERATORS`? That'd match the constructor.

File: pyvrp/__init__.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 from .Population import Population as Population
 from .Population import PopulationParams as PopulationParams
 from .Result import Result as Result
+from .SolveParams import SolveParams as SolveParams
```
Comment (N-Wouda) [inline pyvrp/__init__.py:9]: I think we should also import `solve` here. See also the other comment.

File: tests/test_solve.py
Comment (N-Wouda) [inline tests/test_solve.py:1]: It's a bit tricky to have these tests when they essentially duplicate the tests that also execute `Model.solve` (so essentially everything in `test_Model.py`). That might not be very useful. Some tests that quickly get us much more bang for buck is to focus on a few small smoke tests to check that e.g. running longer/more iters gets a better solution, and that different seeds get different solutions but also that running twice with the same seed gets the same solution. That could be two or three pretty short tests just to check that this should mostly be OK.

**Review by N-Wouda — COMMENTED** [review]
Implementation-wise I think this is essentially OK, but I'm not sure about the place it's currently in, and have some concerns about defaults. The comments below should not take more than say 15min to resolve, I think.

File: pyvrp/cli.py (hunk 3 above)
Comment (leonlan) [reply pyvrp/cli.py]: I find it very useful to have `display` available in the CLI when e.g. solving new VRP variants and see if the search is progressing or not. 

I have added a warning about this when solving a set of instances in parallel. When it's not in parallel, then the output is still OK in the sense that the progress bar doesn't update in place. Does that work for you?

File: pyvrp/solve.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+from pyvrp.crossover import ordered_crossover as ox
+from pyvrp.crossover import selective_route_exchange as srex
+from pyvrp.diversity import broken_pairs_distance as bpd
+from pyvrp.search import LocalSearch, compute_neighbours
+
+if TYPE_CHECKING:
+    import pathlib
+
+    from pyvrp.Result import Result
+    from pyvrp.stop import StoppingCriterion
```
Comment (N-Wouda) [inline pyvrp/solve.py:18]: This can go all the way down to just above `class SolveParams:`, right?

Comment (N-Wouda) [reply pyvrp/solve.py:18]: (The imports below and above seem to be a bit out of whack due to this conditional in the middle)

File: pyvrp/cli.py [hunk 8]
```diff
@@ -222,12 +167,18 @@ def benchmark(instances: list[Path], num_procs: int = 1, **kwargs):
...
     if len(instances) == 1 or num_procs == 1:
         res = [func(arg) for arg in tqdm(args, unit="instance")]
     else:
+        if kwargs.get("display"):
+            warnings.warn(
+                "Solving instances in parallel does not go along well with "
+                "the --display option. Consider disabling this option."
+            )
+
         res = process_map(func, args, max_workers=num_procs, unit="instance")
```
Comment (N-Wouda) [inline pyvrp/cli.py:182]: See Slack!

**Review by N-Wouda — APPROVED** [review]
One small thing about `if TYPE_CHECKING:`

**Pull request merged**